### PR TITLE
Docs: Add guidelines for contributors

### PR DESCRIPTION
**Pull request by mattdot**
This pull request includes updates to the documentation and developer guidelines for the Terraform Provider for Power Platform. The changes aim to provide clearer instructions for contributors and improve the overall structure of the documentation. The most important changes include adding a dynamic configuration guide, enhancing the contribution guidelines, and reorganizing developer resources.

Documentation updates:

* [`.changes/unreleased/documentation-20250303-071315.yaml`](diffhunk://#diff-a4b18a5dc8ca1716060b4d30a20582809f331ca37662c5aa31f8ce4e7e85182aR1-R5): Added a dynamic configuration guide to the user documentation.
* [`CONTRIBUTING.md`](diffhunk://#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055L17-R48): Reorganized and expanded the contribution guidelines, including a new section on examples and clarifying the process for contributing new resources and data sources.

Developer guidelines:

* [`DEVELOPER.md`](diffhunk://#diff-35ea617bfbf0780bfbc554348314c5b11ba478a4d98fd92d33d45cdd54bf326aL7-R155): Reorganized the developer documentation, added a summary of developer resources, and detailed the developer workflow, including setting up development environments using GitHub Codespaces or local devcontainers.
* [`DEVELOPER.md`](diffhunk://#diff-35ea617bfbf0780bfbc554348314c5b11ba478a4d98fd92d33d45cdd54bf326aL112-R216): Updated the section on testing guidelines to provide detailed instructions on writing unit and acceptance tests, and the importance of comprehensive test coverage.This pull request to the Terraform Provider for Power Platform repository includes significant updates to the `CONTRIBUTING.md` and `DEVELOPER.md` files, as well as the addition of new documentation files. The changes aim to improve the contribution guidelines, developer setup instructions, and testing practices. The most important changes include restructuring the contribution types, enhancing developer resource documentation, and adopting `httpmock` for unit testing.

### Contribution Guidelines and Examples:

* [`CONTRIBUTING.md`](diffhunk://#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055L17-R48): Restructured the contribution types section to clarify the difficulty levels of different contributions and added detailed examples for contributing to the Power Platform Terraform Provider.

### Developer Setup and Resources:

* [`DEVELOPER.md`](diffhunk://#diff-35ea617bfbf0780bfbc554348314c5b11ba478a4d98fd92d33d45cdd54bf326aL7-R135): Enhanced the documentation by adding a comprehensive summary of developer resources, including guidelines for testing, security, schema design, and community support. Introduced two recommended development environment options: GitHub Codespaces and Local Devcontainer.
* [`DEVELOPER.md`](diffhunk://#diff-35ea617bfbf0780bfbc554348314c5b11ba478a4d98fd92d33d45cdd54bf326aL90-L94): Provided detailed instructions for setting up and configuring Git credentials in both GitHub Codespaces and Local Devcontainer environments.

### Decision Logs and Testing:

* [`decision-log/loosely_typed_resources.md`](diffhunk://#diff-71472d2aec10121d4e2dfd8ba82a72c9178792f07dcc2283a03b1c466d1fd3beL1-L39): Removed the decision log for loosely typed resources, which outlined the rationale and comparison between strongly typed and loosely typed resources in the Terraform Provider for Power Platform.
* [`devdocs/adr/httpmocks.md`](diffhunk://#diff-693e788c1df485ef4941ea2289e25b2c2dabcdf473d2766c25dc5d848109177cR1-R50): Added a new ADR document detailing the adoption of `httpmock` for HTTP request mocking in unit tests, highlighting the benefits and implementation guidelines for improved test reliability and determinism.This pull request includes several documentation updates and additions to improve clarity and support for developers and users of the Terraform Provider for Power Platform. The most important changes are summarized below:

### Documentation Updates:

* **Decision Log Removal**:
  * Removed the decision log outlining the rationale for implementing loosely typed Terraform resources in `decision-log/loosely_typed_resources.md`.

* **New ADR Document**:
  * Added a new ADR document summarizing the rationale and benefits of loosely typed Terraform resources, specifically `powerplatform_data_record` and `powerplatform_data_records` in `devdocs/adr/loosely_typed_data_records.md`.

* **Community Support Documentation**:
  * Added a comprehensive community support document including FAQs, developer support channels, and best practices for working with maintainers in `devdocs/community_support.md`.

* **Observability and Logging Guidelines**:
  * Introduced guidelines for observability and logging, emphasizing the use of HashiCorp’s `tflog` package for structured logging and the importance of not logging sensitive information in `devdocs/observability_guidelines.md`.

### CODE REVIEW

File: DEVELOPER.md [hunk 1]
```diff
@@ -4,72 +4,155 @@ The Terraform Provider for Power Platform extends Terraform's capabilities to al
...
+1. Add and/or update unit and acceptance tests. Tests for new features should be created in a new resource/datasource_test.go file (commands: `make unittest`, `make acctest`)
+    - When working on a bug, remember to add a new unit and acceptance test(s) covering your use case if that test does not exist yet.
+    - When working on a new feature, add unit and acceptance tests covering the [happy path](https://en.wikipedia.org/wiki/Happy_path) for your feature, ideally also some edge cases. If your feature enhances an existing resource/datasource, add/change validation of your new properties in all tests that use that resource/datasource.
+1. Create/Update examples in `/examples/...` folder(s)
+    - When working on enhancement, remember to add new enhancement properties to all existing examples using that resource/datasource, especially if it is a required property.
+    - When creating a new resource/datasource, create new examples showcasing how to use it.
+1. Regenerate the docs (commands: `make userdocs`)
+1. Create a changelog record for your contribution (commands: `changie new`)
+1. Raise a pull request from your fork back to this repository
```
Comment (mawasile) [inline DEVELOPER.md:42]: 1. or just use `,make precommit`

File: DEVELOPER.md [hunk 2]
```diff
@@ -4,72 +4,155 @@ The Terraform Provider for Power Platform extends Terraform's capabilities to al
...
+## Option 1: GitHub Codespaces (Preferred)
+
+GitHub Codespaces provides a fully managed, cloud-hosted development environment accessible directly from your browser or VS Code. It includes all necessary dependencies and tools pre-installed, allowing you to start contributing immediately.
+
+### Getting Started with Codespaces
+
+1. Navigate to the repository on GitHub.
+2. Click the green **Code** button, then select **Open with Codespaces** > **New codespace**.
+3. GitHub will automatically build and launch your Codespace environment. This may take a few minutes.
+4. Once ready, you can edit, run, debug, and test the code directly in your browser or connect to it from your local VS Code.
```
Comment (mawasile) [inline DEVELOPER.md:65]: do we have to explain how codespace UI works?

File: DEVELOPER.md [hunk 3]
```diff
@@ -4,72 +4,155 @@ The Terraform Provider for Power Platform extends Terraform's capabilities to al
...
-You can setup access to you tenant using the following [bootstrap readme](https://github.com/microsoft/power-platform-terraform-quickstarts/blob/main/bootstrap/README.md) from our quickstarts repository.
+Refer to the [provider's user documentation](https://microsoft.github.io/terraform-provider-power-platform#authentication) for detailed instructions on authenticating to the provider.  When getting started we recommend authenticating with a user-context using `az login` command from az cli and `use_cli` provider configuration.
 
-### Credentials
+#### Service Principals
 
-See the [provider's user documentation](https://microsoft.github.io/terraform-provider-power-platform#authentication) on authenticating to the provider once your tenant is set up.
+While we recommend getting started using user-context authentication, most production scenarios will use service principals. Some APIs behave differently when called in a user-context vs service principal context so we highly recommend testing your contributions with a service principal.
```
Comment (mawasile) [inline DEVELOPER.md:146]: we should add one more sentence, that new feature should be tested not only with user login but also with sp login

File: DEVELOPER.md [hunk 4]
```diff
@@ -109,61 +186,34 @@ docker exec -u vscode -w /workspaces/terraform-provider-power-platform -it <your
...
 
-```bash
-sudo chown -R vscode /workspaces/terraform-provider-power-platform/
-sudo chown -R vscode /go/pkg
-```
+Our testing approach includes both unit tests and acceptance tests:
 
-## Writing Tests
+- **Unit tests**: Fast-executing tests that verify individual components in isolation
+- **Acceptance tests**: End-to-end tests that validate provider behavior against actual Power Platform APIs
```
Comment (mawasile) [inline DEVELOPER.md:196]: we recommend that every acc test has its unit test equivalent. There are exceptions but, that would be a general rule

File: DEVELOPER.md [hunk 5]
```diff
@@ -175,77 +225,68 @@ go get github.com/author/dependency
...
+   - In addition to reading the schema, the tool merges content from the [template files](/templates/). These template files often contain the main structure of the user documentation, placeholders for resource attributes, usage examples, and any provider-level guidance or usage instructions.
+   - If you need to provide custom instructions, additional examples, or overviews, this is done in the templates (rather than in the automatically-generated docs).
 
-Our releases use [semantic versioning](https://semver.org/).
+1. **Examples in `/examples`**
+   - Data Source examples in `examples/data-sources/{your_data_source}/` should have `data-source.tf`, `outputs.tf`, and optionally `variables.tf` files
+   - Resources examples in `examples/resources/{your_data_source}/` should have `resource.tf`, `outputs.tf` files, and optionally `variables.tf` files.  Including a `import.sh` will signal that your resource is compatible with the `import` command.
 
-Given a version number MAJOR.MINOR.PATCH, increment the:
+Before making your commit, open a terminal in your devcontainer and type
```
Comment (mawasile) [inline DEVELOPER.md:258]: type,,,, make precommit :)

File: devdocs/adr/httpmocks.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
+# ADR: Use of httpmock for HTTP Request Mocking in Unit Tests
```
Comment (mawasile) [inline devdocs/adr/httpmocks.md:1]: is there somewhere a guide line, how and where we store req/resp json files?

File: devdocs/testing_guidelines.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+- The test suite includes **pre-checks** (`testAccPreCheck(t)`) to validate required environment variables before running tests.
+- Ensure necessary credentials are set before executing acceptance tests.
+- Missing variables will result in skipped tests to prevent erroneous failures.
+
+#### Running Acceptance Tests
+
+To run all acceptance tests:
+
+```bash
+make acctest
```
Comment (mawasile) [inline devdocs/testing_guidelines.md:69]: I think you need "TF_ACC=1" to get them running? Or we have that set in the container already?

Comment (mattdot) [reply devdocs/testing_guidelines.md:69]: Makefile handles that for unittest

unittest:
	clear
	$(MAKE) clean
	$(MAKE) install
	TF_ACC=0 go test -p 16 -timeout 10m -v ./... -run "^TestUnit$(TEST)"

acctest:
	clear
	$(MAKE) clean
	$(MAKE) install
	TF_ACC=1 go test -p 10 -timeout 300m -v ./... -run "^TestAcc$(TEST)"

File: DEVELOPER.md (hunk 3 above)
Comment (mattdot) [reply DEVELOPER.md:146]: isn't that the next sentence? 😂

File: devdocs/community_support.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+### Q: Where can I find usage examples or documentation for end users?
+
+**A:** End user documentation is available on the [Terraform Registry: Power Platform Provider Docs](https://github.com). It includes detailed resource and data source documentation, examples, and authentication configuration. Additional resources include an [official Microsoft Learn article](https://learn.microsoft.com) and the [QuickStarts repository](https://github.com) containing real-world Terraform configurations.
```
Comment (ianjensenisme) [inline devdocs/community_support.md:15]:
```suggestion
**A:** End user documentation is available on the [Terraform Registry: Power Platform Provider Docs](https://registry.terraform.io/providers/microsoft/power-platform/latest/docs). It includes detailed resource and data source documentation, examples, and authentication configuration. Additional resources include an [official Microsoft Learn article](https://learn.microsoft.com/en-us/business-applications/playbook/enterprise-solutions/power-platform-terraform-provider/) and the [QuickStarts repository](https://github.com/microsoft/power-platform-terraform-quickstarts) containing real-world Terraform configurations.
```

File: devdocs/schema_guidelines.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+Decide carefully if an attribute should be:
+
+- **Required:** For example, `environment_id` and `table_logical_name` in `powerplatform_data_record` are required.
+- **Optional:** Attributes like display names or settings which have default values or can be omitted.
+- **Computed:** Attributes like resource IDs (`id`) or system-generated values, determined by the Power Platform API post-creation.
+- **Optional + Computed:** Used for attributes that can either be specified by the user or defaulted by the system. This pattern is common when the user may optionally provide a value that the system can otherwise determine. For example, certain default configurations or auto-generated fields can follow this approach.
+
+## Naming Attributes
+
+Attribute naming should closely mirror terminology used by the Power Platform API or the Power Platform Admin Center UI. When discrepancies exist due to feature renaming or updates, choose the modern, user-recognizable term. For instance, prefer Dataverse instead of older, deprecated terminology like CDS to ensure users familiar with current platform features easily understand the schema.
```
Comment (ianjensenisme) [inline devdocs/schema_guidelines.md:16]: "modern" and "user-recognizable" are sometimes contradictory. Example: Our Managed Identity references should really be abbreviated "MI" (matching the modern name), but the "user-recognizable" naming could be "MSI". Do we have any guidance to tip the scale in a conflict? In the above example we went with user-recognizable, but I wonder if it's always that clear.

**Review by ianjensenisme — COMMENTED** [review]
Minor updates requested

File: devdocs/schema_guidelines.md (hunk 9 above)
Comment (mattdot) [reply devdocs/schema_guidelines.md:16]: l think it's a judgement call at that point.

File: devdocs/community_support.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+### Q: Where can I find usage examples or documentation for end users?
+
+**A:** End user documentation is available on the [Terraform Registry: Power Platform Provider Docs](https://registry.terraform.io/providers/microsoft/power-platform/latest/docs). It includes detailed resource and data source documentation, examples, and authentication configuration. Additional resources include an [official Microsoft Learn article](https://learn.microsoft.com/en-us/business-applications/playbook/enterprise-solutions/power-platform-terraform-provider/) and the [QuickStarts repository](https://github.com/microsoft/power-platform-terraform-quickstarts) containing real-world Terraform configurations.
+
+### Q: I wrote an acceptance test, but it’s failing due to permissions or timeouts.
+
+**A:** Ensure your test environment has appropriate permissions. For instance, testing environment creation requires admin credentials. Timeouts can be adjusted via Terraform Plugin SDK’s testing framework. Ensure no leftover resources from previous tests are causing conflicts. For persistent issues, open a discussion or issue marked related to testing on GitHub.
+
```
Comment (mawasile) [inline devdocs/community_support.md:20]:
```suggestion
```

File: devdocs/community_support.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
```
Comment (mawasile) [inline devdocs/community_support.md:21]:
```suggestion
```

File: devdocs/schema_guidelines.md [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+Balance strictness and flexibility by validating only what you know to be invalid. Avoid overly constraining attributes to allow flexibility for backend changes that might introduce new valid values without requiring schema updates.
+
+## Structuring Schemas with Nested Attributes (Blocks)
+
+Nested blocks logically group related attributes, improving clarity:
+
+- **SingleNested Blocks:** Used for singular configurations, such as Dataverse settings within the `powerplatform_environment` resource, grouping attributes like `language_code`, `currency_code`, and `security_group_id`.
+- **SetNested or ListNested Blocks:** Used for repeatable structures, such as multiple connector references or multiple similar configurations.
+
+Avoid overly deep nesting. Clearly represent resource or data source structure without unnecessary complexity.
```
Comment (mawasile) [inline devdocs/schema_guidelines.md:63]: I would add more info about judgment, how to structure a new resource. Right now if something has a complex structure like for example env settings, we try to follow the ppac UI in our schema.

**Review by mawasile — DISMISSED** [review]

File: devdocs/release_guidelines.md [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+Using Changie, maintainers (and contributors) include a small changelog snippet for each PR under a directory (by default, `.changes/unreleased`). Each snippet is annotated with the type of change (kind) and the issue/PR number. For instance, a PR that adds a new resource might include an entry marked as “Added” with the PR number. Changie keeps these change files until release time.
+
+**How PRs should be marked:** Contributors should label their pull requests with the appropriate change category or include the corresponding changelog fragment. This ensures the change is documented and that the next release’s version number will be chosen correctly. The `.changie.yaml` config uses these labels to decide how to increment the version. For example, if any PR since the last release was labeled **Breaking**, the next version will be a new major version. If there’s no breaking change but there are new features labeled **Added**, it will be a minor bump; otherwise, for only fixes, it will be a patch bump. Maintainers should enforce this in code review: every PR that changes behavior should either be labeled or accompanied by a changelog entry indicating its type.
+
```
Comment (eduardodfmex) [inline devdocs/release_guidelines.md:37]: Maybe here we can add a comment on Inclusion in Code Review: 

```suggestion
**Inclusion in Code Review** Below are some points which emphasize why inclusivity in code reviews is important:
- Code reviews are an important part of our job as software professionals.
- We work with cross cultural teams from across the globe.
- How we communicate affects team morale.
- Inclusive code reviews welcome new developers and make them comfortable with the team.
- Rude or personal attacks doing code reviews alienate - people can unknowingly make rude comments when reviewing pull requests (PRs).

We consider it important to be inclusive when reviewing PRs. If you want to see how to apply this practice and find some examples, you can check the following link: [Inclusion in Code Review](https://microsoft.github.io/code-with-engineering-playbook/code-reviews/inclusion-in-code-review/)

```

**Review by eduardodfmex — COMMENTED** [review]
Thank you for creating and extending the documentation..! 
I add a couple of suggestions.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -14,37 +14,35 @@ contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additio
...
-- Ensure the PR description clearly describes the feature you're adding and any known limitations
-
-## Decision Log
-
-The decision log is a record of significant design decisions made during the development of the Terraform Provider for Power Platform. It provides historical context and rationale for these decisions, helping contributors and maintainers understand the reasoning behind certain design choices.  Inspirations for the design decision log are [Microsoft's Engineering Playbook Decision Logs](https://microsoft.github.io/code-with-engineering-playbook/design/design-reviews/decision-log/) and [Hashicorp's Plugin Framework design log](https://github.com/hashicorp/terraform-plugin-framework/tree/main/docs/design).
+- [Update auto-generated documentation](./DEVELOPER.md#updating-documentation) in [/docs](/docs/). (Don't manually edit [/docs](/docs/) or your updates will be overwritten)
+- Ensure the PR description clearly describes the feature you're adding and any known limitations (We recommend using GitHub Copilot for PR descriptions)
```
Comment (eduardodfmex) [inline CONTRIBUTING.md:45]: Maybe here we can add a comment on Inclusion in Code Review:
```suggestion

## Inclusion in Code Review
Below are some points which emphasize why inclusivity in code reviews is important:
- Code reviews are an important part of our job as software professionals.
- We work with cross cultural teams from across the globe.
- How we communicate affects team morale.
- Inclusive code reviews welcome new developers and make them comfortable with the team.
- Rude or personal attacks doing code reviews alienate - people can unknowingly make rude comments when reviewing pull requests (PRs).
We consider it important to be inclusive when reviewing PRs. If you want to see how to apply this practice and find some examples, you can check the following link: [Inclusion in Code Review](https://microsoft.github.io/code-with-engineering-playbook/code-reviews/inclusion-in-code-review/)

```

File: devdocs/community_support.md [hunk 15]
```diff
@@ -0,0 +1,49 @@
+# Community and Support Documentation
+
+## FAQ (Frequently Asked Questions)
+
+### Q: Can I develop on this provider without using the VS Code Dev Container?
+
+**A:** It's not recommended to develop this provider outside the provided Dev Container. GitHub Codespaces is the preferred environment due to consistency, controlled dependencies, and reproducibility. Using the Dev Container locally is acceptable, but manually setting up a local environment can introduce risks such as mismatched dependencies, version conflicts, and environmental inconsistencies that could complicate development and troubleshooting.
```
[the remark this answers is not in the document]
Comment (mattdot) [reply devdocs/community_support.md:7]: the last part of the sentence is supposed to be about not using the devcontainer.  It says devcontainer locally is fine.

File: devdocs/testing_guidelines.md [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+```
+
+To run a single acceptance test:
+
+```bash
+TF_ACC=1 go test -v ./... -run TestAcc<test_name>
+```
+
```
[the remark this answers is not in the document]
Comment (mattdot) [reply devdocs/testing_guidelines.md:77]: i think we should change the actual timeouts in makefile

File: devdocs/community_support.md (hunk 10 above)
Comment (mattdot) [reply devdocs/community_support.md:20]: will just change the make to have a longer timeout.

**Review by leighatami — APPROVED** [review]
